### ISSUE TITLE
Status of the Volume of Fluid implementation

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#3186]
In one of those bizarre turns at workshops, I spent twenty minutes talking to someone from a completely different field about the volume of fluid implementation in ASPECT today. This also made me think about the current status of that work.

In particular:
* In #835, there was talk about a cookbook that would explain how to use the VoF feature
* If I recall correctly, the VoF implementation does not completely (or at all?) work in the case of adaptively refined meshes. For example, there is this assertion:
```
              // We are currently assembling for a case where the AMR refinement level differs.
              // If the volume fraction of the fluid flux is not 0 or 1 (to
              // within tolerance), we are at a VoF fluid interface but have
              // not correctly computed the advection flux
              AssertThrow(flux_volume_of_fluid < volume_fraction_threshold || flux_volume_of_fluid>1.0-volume_fraction_threshold,
                          ExcMessage("Volume of Fluid interface coincides with variation in AMR refinement levels. "
                                     "This case is not handled by the current Volume of Fluid assembler."));
```
* I recall that there were also issues in 3d. Indeed, all existing tests only run in 2d.

With these limitations, it seems unclear to me that this feature will eventually be very useful to a broader community -- which was, of course, the goal of the SI2 program that funded the VoF development :-)

@egpuckett, @class4kayaker: Do you have a plan to address these things? How should we move forward with the VoF implementation?

In reference to #835.

**Comment by class4kayaker** [comment]
- I have not yet done a cookbook for the VoF implementation due in part to @egpuckett  indicating some interest in writing that, while I also had my dissertation to write up.
- For AMR meshes, assuming that the actual interface and the surrounding cells are always at a consistent refinement level there is no problem. I did find a case where  The `volume of fluid interface` refinement strategy plugin enforces this, and is currently explicitly required for AMR meshes when VoF is in use. The assertion you note was included due to being a case that might show up due to errors somewhere, that violates some of the necessary assumptions (think of cases like density<0, they shouldn't happen, but some edge cases might force them and we need to generate an error in such cases). PR #3174 dealt with a case where the refinement strategy was failing to enforce this requirement due to me forgetting about some edge cases with multiple processes.
- The algorithm is currently limited to 2d due to the reconstruction algorithm used, and the limitations on the current functions to calculate enclosed volume from interface location. Both of these should accommodate extension to 3d under the current design, but have a number of edge cases that make implementation more involved.

**Comment by bangerth** [comment]
@egpuckett and I just had a long conversation. Here's what we resolved:
* Gerry has a bunch of models that show the VoF, which he'll send to me, and we'll pick something that would be useful for a cookbook.
* He'll write the cookbook based on that, showing how to use the VoF method.
* I'm volunteering to review it.

This at least addresses the lack of a cookbook showing the method. That leaves two issues:
* The fact that the reconstructed interface has to be entirely on the finesh mesh level.
* The fact that the VoF method is not implemented in 3d.
Absent finding funding for @class4kayaker, we don't have a solution for these.

**Comment by bangerth** [comment]
@egpuckett -- have you had a chance to think about the cookbook some more?

**Comment by egpuckett** [comment]
Yes ... In fact Jon has made an excellent draft with superb computations and I have edited it. Now he has added some more edits and data and it is sitting in my inbox.  I am the bottleneck.  (See below.)

I want to say that the results compare extremely favorably on the van Keken problem when one uses compositional fields and I am eager to have everyone who wrote the van Keken pages see it.

I don't know whether to take the remainder of what I am going to write here offline or not. You can let me know about future such remarks. The reason I haven't moved faster on this is me and the doubled or perhaps tripled workload associated with CV19.  

The city of Davis is under a shelter in place order.  No one has been allowed on campus since a week ago or so and no nonessential personnel are allowed on until at least April 7.  There is no guarantee that we'll be able to return to campus then.  What is certain is that starting March 30 we will begin the Spring Quarter (SQ) completely online.  Everyone at UCD who is teaching.

This is finals week for our Winter Quarter.  We could not have an in person final. My TA and I just created and gave an online final on Tuesday 6:00 - 8:00 PM to 67 students.  My TA is running fever and I am going to contact him and suggest he seek medical help.  There are at least 67 additional Final Programming Project papers he was going to grade. Then I have to grade another problem and assign final grades. (There's more, but I won't go into detail here.)

I just got out of a two hour Zoom meeting with my colleagues in the Math Department discussing the VERY large number of options for delivering instruction online for SQ which starts on Monday, March 30.  The amount of work to prepare to teach a class of 95 students by then is enormous.  I've had two Zoom meetings today and another two hour online exam tomorrow morning.

Anyway, I am the bottleneck, with respect to the Cookbook.  I really want one last pass at the Cookbook before Jon makes a PR. If you want to see it Jon or I can give you access to the branch it's on.  (In fact it may be public on my GitHub page

[VOF_VAN_KEKEN_COOKBOOK](https://github.com/EGP-CIG-REU/aspect/tree/VOF_VAN_KEKEN_COOKBOOK)

If you want that branch, let me know.  I have to pull Hon's branch and then push his new material to my branch.

However, I would REALLY like to make my final best effort and then do the PR. I'm happy to have you look at it now, but I would like you to wait to make comments until I'm finished with my final edits and then we make the PR. I was hoping to do it this morning, but things just keep changing so quickly here.  

I hope to have it done in a few days.  Perhaps even tomorrow, but I can't make promises.  Life keeps changing. :-) 

Sorry for what might sound like a bunch of excuses.  

There is nothing I would treasure more right now than to have a beer with you and all of our ASPECT colleagues!

I hope everyone and their loved ones are safe and healthy!

**Comment by egpuckett** [comment]
@bangerth 

    [VOF_VAN_KEKEN_COOKBOOK](https://github.com/EGP-CIG-REU/aspect/tree/VOF_VAN_KEKEN_COOKBOOK)

is now current with Jon's branch.  I am planning to edit some now.  No promises when I finish.  My students are  in constant contact for various reasons.

**Comment by bangerth** [comment]
@egpuckett -- no worries. I don't need it today :-) All I needed to know is that *something* is happening, there is no urgency is getting it reviewed and commented on *today* :-))

**Comment by egpuckett** [comment]
@bangerth Well you got it today. :-)  The testers won't run, but do you need them to.  Anyway I look forward to you thoughts on the quality of the results.

Best Wishes,

- G

**Comment by bangerth** [comment]
@egpuckett -- are you asking me to review the current state? Could you make it into a regular pull request?

**Comment by egpuckett** [comment]
I asked Jon to make it a regular PR last week, or whatever 4 days ago was.  It's PR#3425.

It didn't pass the test at first.  (It is apparently still not passing the tests, I don't know why.) Timo made some comment.  Jon wrote to me 

> Looking at the case, I saw that it was a case of the tests not having
permission to run. While checking I noted that there was a modification
outside of the edits to the manual, so I deleted that, and did a rebase.

and then

> We need someone with permissions to authorize running the tests, but
given that we do not touch the actual source code, they may decide not
to run them.

I gather we can't 'assign' a person (e.g. you) review our PR.  I was not sure how to let you know we had made it.  I guess mentioning it in this Issue / Discussion didn't work.

Well no harm done.  I just filed final grades for last quarter, so it would have been difficult to participate in a Review Discussion before now.

Now all I have to do is figure out how to deliver instruction online via Zoom to 95 students starting Monday.  If you have any suggestions, I would be grateful for them.  (It's Sophomore Linear Algebra.)

**Comment by egpuckett** [comment]
@bangerth The VOF cookbook #3425 has been reviewed by @naliboff  He and @class4kayaker are ready to merge. I asked that they wait until I double check what changes John Naliboff made to the cookbook in the ASPECT Manual, since I wrote the final edit of the cookbook and want to see what is different.  (John says he only made minor edits, like commas and that sort of thing.)  I'm part way through and I have promised myself to finish before the end of the Hackathon and give John Naliboff my blessings to merge it.  I suspect you are too busy right now to read it. However, I would be grateful if at some point you would find the time to do that. I think the results compare very favorably with the other van Keken benchmark results in the Manual.

A quick note on a related issue.  @anne-glerum mentioned the instabilities with the free-surface algorithm that are being damped by adding diffusion. I think the VOF method should be able to ameliorate the instabilities, since it was developed at LLNL to work in ALE codes there and then used by Hirt and Nichols at LANL to model free-surfaces.  When Anne has the time I plan to look at this with her and work on it in 2D by asking @class4kayaker questions when I need help. I am committed to getting VOF into 3D and curvilinear grids and coming up to speed so that I can support it.  This will take some time, but I promise the ASPECT community will benefit from having VOF in ASPECT.
